bindings/go/dag/dag.go

```diff
@@ -0,0 +1,328 @@
+
+	delete(d.Vertices, id)
+	delete(d.OutDegree, id)
+	delete(d.InDegree, id)
+
+	return nil
+}
+
+type CycleError struct {
+	Cycle []string
+}
+
+func (e *CycleError) Error() string {
+	return fmt.Sprintf("The current graph would create a cycle: %s", formatCycle(e.Cycle))
+}
+
+func formatCycle(cycle []string) string {
+	return strings.Join(cycle, " -> ")
+}
+
+// AddEdge adds a directed edge from one node to another.
+func (d *DirectedAcyclicGraph[T]) AddEdge(from, to T, attributes ...map[string]any) error {
+	fromNode, fromExists := d.Vertices[from]
+	_, toExists := d.Vertices[to]
+	if !fromExists {
+		return fmt.Errorf("node %v does not exist", from)
+	}
+	if !toExists {
+		return fmt.Errorf("node %v does not exist", to)
+	}
+	if from == to {
+		return ErrSelfReference
+	}
+
+	_, exists := fromNode.Edges[to]
+
+	if !exists {
+		// Only initialize the map if the edge was added
+		fromNode.Edges[to] = map[string]any{}
+		// Only increment the out-degree and in-degree if the edge was added
+		d.OutDegree[from]++
+		d.InDegree[to]++
+
+		// Check if the graph is still a DAG
+		hasCycle, cycle := d.HasCycle()
+		if hasCycle {
+			// Ehmmm, we have a cycle, let's remove the edge we just added
+			delete(fromNode.Edges, to)
+			d.OutDegree[from]--
+			d.InDegree[to]--
+
+			return fmt.Errorf("adding an edge from %v to %v would create a cycle: %w", fmt.Sprintf("%v", from), fmt.Sprintf("%v", to), &CycleError{
+				Cycle: cycle,
+			})
+		}
+	}
+
+	for _, attributes := range attributes {
+		maps.Copy(fromNode.Edges[to], attributes)
+	}
+
+	return nil
+}
+
+func (d *DirectedAcyclicGraph[T]) Roots() []T {
+	var roots []T
+	for node, degree := range d.InDegree {
+		if degree == 0 {
+			roots = append(roots, node)
+		}
+	}
+	return roots
+}
+
+func (d *DirectedAcyclicGraph[T]) TopologicalSort() ([]T, error) {
+	if cyclic, nodes := d.HasCycle(); cyclic {
+		return nil, &CycleError{
+			Cycle: nodes,
+		}
+	}
+
+	visited := make(map[T]bool)
+	var order []T
+
+	// Get a sorted list of all vertices
+	vertices := d.GetVertices()
+
+	var dfs func(T)
+	dfs = func(node T) {
+		visited[node] = true
+
+		// Sort the neighbors to ensure deterministic order
+		neighbors := make([]T, 0, len(d.Vertices[node].Edges))
+		for neighbor := range d.Vertices[node].Edges {
+			neighbors = append(neighbors, neighbor)
+		}
+		slices.Sort(neighbors)
+
+		for _, neighbor := range neighbors {
+			if !visited[neighbor] {
+				dfs(neighbor)
+			}
+		}
+		order = append(order, node)
+	}
+
+	// Visit nodes in a deterministic order
+	for _, node := range vertices {
+		if !visited[node] {
+			dfs(node)
+		}
+	}
+
+	return order, nil
+}
+
+// GetVertices returns the nodes in the graph in sorted alphabetical
+// order.
+func (d *DirectedAcyclicGraph[T]) GetVertices() []T {
+	nodes := make([]T, 0, len(d.Vertices))
+	for node := range d.Vertices {
+		nodes = append(nodes, node)
+	}
+
+	// Ensure deterministic order. This is important for TopologicalSort
+	// to return a deterministic result.
```

Comment:
If we care so much about order, I wonder why not just have them as Slices and have a map as a side tracker for IDs if really need it. If we end up sorting during all GetVertices call which is quite plenty, this could be super bad for performance.

Reply:
Though I would have to run a few benchmarks first to be sure about this statement.

Reply:
We care about edge ordering, not necessarily slice ordering. Thats quite different

Reply:
So you're saying that even if the slice is ordered or rather created in some order it doesn't mean that a slice would help vs a map since you'll have to order it anyways.